File: backend-ts/src/api/schemas/assignment.schema.ts
import { z } from 'zod';

/** Programming language enum */
export const ProgrammingLanguageSchema = z.enum(['python', 'java', 'c']);
export type ProgrammingLanguage = z.infer<typeof ProgrammingLanguageSchema>;

/** Create assignment request schema */
export const CreateAssignmentRequestSchema = z.object({
    teacherId: z.number().int().min(1),
    assignmentName: z.string().min(1).max(150),
    description: z.string().min(1),
    programmingLanguage: ProgrammingLanguageSchema,
    deadline: z.string().datetime(),
    allowResubmission: z.boolean().default(true),
    maxAttempts: z.number().int().min(1).max(99).nullable().optional(),
    templateCode: z.string().max(50000).nullable().optional(),
});

export type CreateAssignmentRequest = z.infer<typeof CreateAssignmentRequestSchema>;

/** Update assignment request schema */
export const UpdateAssignmentRequestSchema = z.object({
    teacherId: z.number().int().min(1),
    assignmentName: z.string().min(1).max(150).optional(),
    description: z.string().min(1).optional(),
    programmingLanguage: ProgrammingLanguageSchema.optional(),
    deadline: z.string().datetime().optional(),
    allowResubmission: z.boolean().optional(),
    maxAttempts: z.number().int().min(1).max(99).nullable().optional(),
    templateCode: z.string().max(50000).nullable().optional(),
});

export type UpdateAssignmentRequest = z.infer<typeof UpdateAssignmentRequestSchema>;

/** Assignment response schema */
export const AssignmentResponseSchema = z.object({
    id: z.number(),
    classId: z.number(),
    assignmentName: z.string(),
    description: z.string(),
    programmingLanguage: z.string(),
    deadline: z.string(),
    allowResubmission: z.boolean(),
    maxAttempts: z.number().nullable().optional(),
    createdAt: z.string(),
    isActive: z.boolean(),
    templateCode: z.string().nullable().optional(),
    hasTemplateCode: z.boolean().optional(),
    submissionCount: z.number().optional(),
    hasSubmitted: z.boolean().optional(),
});

export type AssignmentResponse = z.infer<typeof AssignmentResponseSchema>;

/** Assignment detail response schema */
export const AssignmentDetailResponseSchema = AssignmentResponseSchema.extend({
    className: z.string().optional(),
    teacherName: z.string().optional(),
});

export type AssignmentDetailResponse = z.infer<typeof AssignmentDetailResponseSchema>;

// ============================================================================
// Param & Query Schemas
// ============================================================================

/** Assignment ID param schema */
export const AssignmentIdParamSchema = z.object({
    assignmentId: z.string(),
});

export type AssignmentIdParam = z.infer<typeof AssignmentIdParamSchema>;

// ============================================================================
// Response Schemas (for controller routes)
// ============================================================================

/** Get assignment response schema */
export const GetAssignmentResponseSchema = z.object({
    success: z.literal(true),
    message: z.string(),
    assignment: AssignmentDetailResponseSchema,
});

export type GetAssignmentResponse = z.infer<typeof GetAssignmentResponseSchema>;

/** Update assignment response schema */
export const UpdateAssignmentResponseSchema = z.object({
    success: z.literal(true),
    message: z.string(),
    assignment: AssignmentResponseSchema,
});

export type UpdateAssignmentResponse = z.infer<typeof UpdateAssignmentResponseSchema>;

/** Assignment list response schema */
export const AssignmentListResponseSchema = z.object({
    success: z.literal(true),
    message: z.string(),
    assignments: z.array(AssignmentResponseSchema),
});

export type AssignmentListResponse = z.infer<typeof AssignmentListResponseSchema>;

/** Create assignment response schema */
export const CreateAssignmentResponseSchema = z.object({
    success: z.literal(true),
    message: z.string(),
    assignment: AssignmentResponseSchema,
});

export type CreateAssignmentResponse = z.infer<typeof CreateAssignmentResponseSchema>;

File: backend-python-deprecated/api/schemas/assignment_schema.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, Literal
from datetime import datetime, timezone


class CreateAssignmentRequest(BaseModel):
    """Request schema for creating an assignment"""
    class_id: int = Field(..., description="ID of the class")
    teacher_id: int = Field(..., description="ID of the teacher (for authorization)")
    assignment_name: str = Field(..., min_length=1, max_length=150, description="Name of the assignment")
    description: str = Field(..., min_length=1, description="Assignment description")
    programming_language: Literal["python", "java"] = Field(..., description="Programming language")
    deadline: datetime = Field(..., description="Assignment deadline")
    allow_resubmission: bool = Field(default=True, description="Whether to allow resubmissions")

    @field_validator('assignment_name')
    @classmethod
    def validate_assignment_name(cls, v: str) -> str:
        """Validate and trim assignment name"""
        trimmed = v.strip()
        if not trimmed:
            raise ValueError('Assignment name cannot be empty')
        if len(trimmed) > 150:
            raise ValueError('Assignment name must not exceed 150 characters')
        return trimmed

    @field_validator('description')
    @classmethod
    def validate_description(cls, v: str) -> str:
        """Validate and trim description"""
        trimmed = v.strip()
        if not trimmed:
            raise ValueError('Description cannot be empty')
        if len(trimmed) < 10:
            raise ValueError('Description must be at least 10 characters')
        return trimmed

    @field_validator('deadline')
    @classmethod
    def validate_deadline(cls, v: datetime) -> datetime:
        """Validate deadline is in the future"""
        # Always use UTC for comparison to match database timezone
        now = datetime.now(timezone.utc)

        # If v is naive, make it UTC-aware
        if v.tzinfo is None:
            v = v.replace(tzinfo=timezone.utc)

        if v <= now:
            raise ValueError('Deadline must be in the future')
        return v


class UpdateAssignmentRequest(BaseModel):
    """Request schema for updating an assignment"""
    teacher_id: int = Field(..., description="ID of the teacher (for authorization)")
    assignment_name: Optional[str] = Field(None, min_length=1, max_length=150, description="Name of the assignment")
    description: Optional[str] = Field(None, min_length=1, description="Assignment description")
    programming_language: Optional[Literal["python", "java"]] = Field(None, description="Programming language")
    deadline: Optional[datetime] = Field(None, description="Assignment deadline")
    allow_resubmission: Optional[bool] = Field(None, description="Whether to allow resubmissions")

    @field_validator('assignment_name')
    @classmethod
    def validate_assignment_name(cls, v: Optional[str]) -> Optional[str]:
        """Validate and trim assignment name"""
        if v is None:
            return v
        trimmed = v.strip()
        if not trimmed:
            raise ValueError('Assignment name cannot be empty')
        if len(trimmed) > 150:
            raise ValueError('Assignment name must not exceed 150 characters')
        return trimmed

    @field_validator('description')
    @classmethod
    def validate_description(cls, v: Optional[str]) -> Optional[str]:
        """Validate and trim description"""
        if v is None:
            return v
        trimmed = v.strip()
        if not trimmed:
            raise ValueError('Description cannot be empty')
        if len(trimmed) < 10:
            raise ValueError('Description must be at least 10 characters')
        return trimmed

    @field_validator('deadline')
    @classmethod
    def validate_deadline(cls, v: Optional[datetime]) -> Optional[datetime]:
        """Validate deadline is in the future"""
        if v is None:
            return v

        # Always use UTC for comparison to match database timezone
        now = datetime.now(timezone.utc)

        # If v is naive, make it UTC-aware
        if v.tzinfo is None:
            v = v.replace(tzinfo=timezone.utc)

        if v <= now:
            raise ValueError('Deadline must be in the future')
        return v


class AssignmentDetailResponse(BaseModel):
    """Response schema for assignment details"""
    id: int
    class_id: int
    title: str
    description: str
    programming_language: str
    deadline: str
    allow_resubmission: bool
    is_active: bool
    class_name: Optional[str] = None
    created_at: Optional[str] = None

    class Config:
        from_attributes = True
        populate_by_name = True


class CreateAssignmentResponse(BaseModel):
    """Response schema for assignment creation"""
    success: bool
    message: Optional[str] = None
    assignment: Optional[AssignmentDetailResponse] = None


class UpdateAssignmentResponse(BaseModel):
    """Response schema for assignment update"""
    success: bool
    message: Optional[str] = None
    assignment: Optional[AssignmentDetailResponse] = None


class DeleteAssignmentResponse(BaseModel):
    """Response schema for assignment deletion"""
    success: bool
    message: Optional[str] = None


class GetAssignmentResponse(BaseModel):
    """Response schema for getting assignment details"""
    success: bool
    message: Optional[str] = None
    assignment: Optional[AssignmentDetailResponse] = None
